src/gda/backend_gda.cpp

```diff
@@ -1099,123 +1102,18 @@ void GDABackend::create_cqs(int cqe) {
-
-  strncpy(gpu_qp->dev_name,
-          qps[conn_num]->context->device->name,
-          sizeof(gpu_qp->dev_name));
-  gpu_qp->dev_name[sizeof(gpu_qp->dev_name) - 1] = 0;
-
-  gpu_qp->qp_num = qps[conn_num]->qp_num;
-  gpu_qp->lkey = heap_mr->lkey;
-  gpu_qp->rkey = heap_rkey[conn_num % num_pes];
-  gpu_qp->inline_threshold = 32;
-#endif /* GDA_IONIC */
-  if (gda_vendor == GDAVendor::MLX5) {
-    mlx5dv_cq cq_out;
-    mlx5dv_obj mlx_obj;
-    mlx_obj.cq.in = cqs[conn_num];
-    mlx_obj.cq.out = &cq_out;
-    mlx5dv_ftable_.init_obj(&mlx_obj, MLX5DV_OBJ_CQ);
-    dump_mlx5dv_cq(&cq_out, conn_num);
-
-    /*
-     * struct mlx5dv_cq {
-     *   void                    *buf;
-     *   __be32                  *dbrec;
-     *   uint32_t                cqe_cnt;
-     *   uint32_t                cqe_size;
-     *   void                    *cq_uar;
-     *   uint32_t                cqn;
-     *   uint64_t                comp_mask;
-     * };
-    */
-
-    gpu_qp->cq_buf = reinterpret_cast<mlx5_cqe64*>(cq_out.buf);
-    gpu_qp->cq_cnt = cq_out.cqe_cnt;
-    gpu_qp->cq_log_cnt = log2(cq_out.cqe_cnt);
-    gpu_qp->cq_dbrec = cq_out.dbrec;
-
-    mlx5dv_qp qp_out;
-    mlx_obj.qp.in = qps[conn_num];
-    mlx_obj.qp.out = &qp_out;
-    mlx5dv_ftable_.init_obj(&mlx_obj, MLX5DV_OBJ_QP);
-    dump_mlx5dv_qp(&qp_out, conn_num);
-
-    /*
-     * struct mlx5dv_qp {
-     *   __be32 *dbrec;
-     *   struct {
-     *     void *buf;
-     *     uint32_t wqe_cnt;
-     *     uint32_t stride;
-     *   } sq;
-     *   struct {
-     *     void *buf;
-     *     uint32_t wqe_cnt;
-     *     uint32_t stride;
-     *   } rq;
-     *   struct {
-     *     void *reg;
-     *     uint32_t size;
-     *   } bf;
-     *   uint64_t comp_mask;
-     *   off_t uar_mmap_offset;
-     *   uint32_t tirn;
-     *   uint32_t tisn;
-     *   uint32_t rqn;
-     *   uint32_t sqn;
-     *   uint64_t tir_icm_addr;
-     * };
-     */
-
-    gpu_qp->dbrec = &qp_out.dbrec[1]; // points to two pointers: 0 -> MLX5_REC_DBR, 1 -> MLX5_SND_DBR
-    gpu_qp->sq_buf = reinterpret_cast<uint64_t*>(qp_out.sq.buf);
-    gpu_qp->sq_wqe_cnt = qp_out.sq.wqe_cnt;
-    gpu_qp->rkey = htobe32(heap_rkey[conn_num % num_pes]);
-    gpu_qp->lkey = htobe32(heap_mr->lkey);
-    gpu_qp->qp_num = qps[conn_num]->qp_num;
-    gpu_qp->inline_threshold = inline_threshold;
-    // The 2 in qp_out.bf.size * 2 below facilitates the switching between blue flame registers
-    void* gpu_ptr{nullptr};
-    rocm_memory_lock_to_fine_grain(qp_out.bf.reg, qp_out.bf.size * 2, &gpu_ptr, hip_dev_id);
-    gpu_qp->db.ptr = reinterpret_cast<uint64_t*>(gpu_ptr);
+  switch (gda_vendor) {
+  case GDAVendor::IONIC:
+    ionic_initialize_gpu_qp(gpu_qp, conn_num);
+    break;
+  case GDAVendor::BNXT:
+    bnxt_initialize_gpu_qp(gpu_qp, conn_num);
+    break;
+  case GDAVendor::MLX5:
+    mlx5_initialize_gpu_qp(gpu_qp, conn_num);
+    break;
+  default:
+    assert(false /* GDAVendor initialize_gpu_qp */);
```

Comment:
```suggestion
  switch (gda_vendor) {
  case GDAVendor::NONE:
    assert(false /* GDAVendor initialize_gpu_qp */);
#ifdef NDEBUG
    __builtin_unreachable();
#endif
    break;
  case GDAVendor::IONIC:
    ionic_initialize_gpu_qp(gpu_qp, conn_num);
    break;
  case GDAVendor::BNXT:
    bnxt_initialize_gpu_qp(gpu_qp, conn_num);
    break;
  case GDAVendor::MLX5:
    mlx5_initialize_gpu_qp(gpu_qp, conn_num);
    break;
```
Could do something like this and use `-Werror=switch-enum` to catch non-exhaustive switches - useful if we add other GDA vendors in the future. By the time we get here, `gda_vendor` should never be `NONE`.

Reply:
Its a good idea for a cleanup pr later when we are less rushed